src/creature_tracker.h

```diff
@@ -118,20 +133,99 @@ class creature_tracker
-        /** Remove the monsters entry in @ref monsters_by_location */
-        void remove_from_location_map( const monster &critter );
+
+        /**
+         * Creatures that get removed via @ref remove are stored here until the end of the turn.
+         * This keeps the objects valid and they can still be accessed instead of causing UB.
+         */
+        std::unordered_set<shared_ptr_fast<monster>> removed_this_turn_;  // NOLINT(cata-serialize)
+
+        // Tracks the dirtiness of the visitable zones cache. This must be flipped when
+        // persistent visibility from terrain or furniture changes (this excludes vehicles and fields)
+        // or when persistent traversability changes, which means walls and floors.
+        bool dirty_ = true;  // NOLINT(cata-serialize)
+        int zone_tick_ = 1;  // NOLINT(cata-serialize)
+        int zone_number_ = 0;  // NOLINT(cata-serialize)
+        std::unordered_map < int, std::unordered_map<mfaction_id, std::vector<Creature *>>>
+        creatures_by_zone_and_faction_;  // NOLINT(cata-serialize)
+        std::unordered_set<Creature *> removed_;  // NOLINT(cata-serialize)
+
+        friend game;
 };
 
 creature_tracker &get_creature_tracker();
 
+// Implementation Details
+
+template <typename PredicateFn>
+Creature *creature_tracker::find_reachable( const Creature &origin, PredicateFn predicate_fn )
+{
+    return find_reachable( origin, []( const mfaction_id & ) {
+        return true;
+    }, std::move( predicate_fn ) );
+}
+
+template <typename FactionPredicateFn, typename CreaturePredicateFn>
+Creature *creature_tracker::find_reachable( const Creature &origin, FactionPredicateFn faction_fn,
+        CreaturePredicateFn creature_fn )
+{
+    flood_fill_zone( origin );
+
+    const auto map_iter = creatures_by_zone_and_faction_.find( origin.get_reachable_zone() );
+    if( map_iter != creatures_by_zone_and_faction_.end() ) {
+        for( const auto& [faction, creatures] : map_iter->second ) {
+            if( !faction_fn( faction ) ) {
+                continue;
+            }
+            for( Creature *other : creatures ) {
+                if( removed_.count( other ) == 0 ) {
+                    if( creature_fn( other ) ) {
+                        return other;
+                    }
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+template <typename VisitFn>
+void creature_tracker::for_each_reachable( const Creature &origin, VisitFn visit_fn )
+{
+    find_reachable( origin, [visit_fn = std::move( visit_fn )]( Creature * other ) {
+        visit_fn( other );
+        return false;
+    } );
+}
+
+template <typename FactionPredicateFn, typename CreatureVisitFn>
+void creature_tracker::for_each_reachable( const Creature &origin, FactionPredicateFn faction_fn,
+        CreatureVisitFn creature_fn )
+{
+    find_reachable( origin, std::move( faction_fn ), [creature_fn = std::move( creature_fn )](
```

Comment:
What does move acconplish here?

Reply:
The functors can be any object that is invokable. This means it could be a potentially very large object that would be expensive to copy, or the object could by non-copyable entirely, etc the following is valid:

```struct NonCopyableFunctor {
    template <typename... Ts>
    void operator()(Ts&&...) const {}

    std::unique_ptr<int> non_copyable;
};
...
tracker.for_each_reachable(critter, NonCopyableFunctor());
```

By moving everything into place we avoid expensive copies, and allow non-copyable functors.

Reply:
You almost got it right, except the declared type means you're forcing a copy when the function is invoked, which also prevents compiler optimizations.

Every functor should be annotated with && to use universal reference matching and std::forward<>ed along.

Reply:
It's exactly right. It doesn't force a copy. You can try the above and see it compiles just fine. You are free to move the value into the function call, or also just pass a temporary (which is extremely common when using these with lambdas).

I don't use forwarding references for passing in functions specifically, as I've encountered situations in the past where the generated code is complete garbage. You're still essentially passing in a pointer, with all the associated problems that has with aliasing (especially when it comes to captured variables).

Reply:
Here is sort of a minimal example of where using forwarding references with function objects results in worse code: https://godbolt.org/z/qxTq84b3q

It's a little contrived because compilers are smart, so simple programs don't tend to have problems, but I've seen this in production code with significantly less attribute abuse.

Reply:
Aren't you `std::move()`ing to a copy constructor? This is closer to what you have here and the code looks marginally worse with `std::move()` at the call site https://godbolt.org/z/MY8r88d6K

Reply:
Moving into a copy constructor? No? I'm not sure what you're getting at, perhaps it's better to move this conversation into the Discord.

Reply:
The reason I asked is that std::move is abused a lot, and it can break things,  so I don't want it used without a compelling rationale,  which this is not.
Please ease up on the reflexive microoptimization.

Reply:
It wasn't done as a "microoptimization". It was done because it's correct the correct manner to handle passed-by-value template parameters: you either pass by value and move along, or pass by forwarding reference and forward along. Unless of course you know you are always working with primitive types, which doesn't apply in this case.

I'm fine with switching to forwarding references if that's what you want (and let me know), but you should know that it pretty much always results in worse code and bloated binary sizes.

Reply:
I'm aware that this will copy the lambda if the API is used that way, the point is that it isn't a _requirement_ to use the API. As you note, you can move into it, or just pass in a temporary. The primary value add of passing by value is that the function then owns the data passed in, and can do specific optimizations that it cannot do otherwise.

You do make a good point that it's a very common pattern in this codebase for a lambda to _not_ be moved into a function like this. I think the fact that the codebase is worked on by novice programmers, who would likely forget such a thing, would be a good enough reason to switch this to forwarding references.

I'll just swap this to forwarding references, since it is less likely a novice will miss use the API.

> Extraordinary claims require extraordinary evidence, but that discussion is not relevant to this PR.

As for evidence, while I cannot provide it directly, https://www.vldb.org/pvldb/vol12/p2022-chattopadhyay.pdf is a paper I am co-author on that describes a template metaprogramming based SQL evaluation engine. Removing forwarding references for lambdas in the very performance sensitive parts of the code resulted in very noticeable gains. I can discuss the specifics in another place if you're interested, but some of the details may escape me since it was a while ago.